bridgebot/bidding.py

```diff
@@ -1,21 +1,123 @@
     pass
 
+class InvalidBidException(Exception):
+    pass
 
 class Contract:
     def __init__(self):
-        self.level = 0
-        self.strain = Strains.NT
+        self.contract = None
         self.doubled = Doubles.NONE
         self.declarer = None
+        self.last_bid_by = None # Redundant, but cleans up edge case logic in Bids.is_sufficient_bid
+
+
+# Compares to a contract, so can't go into enum
+class Bids(Enum):
+    ONE_CLUB = Contracts.ONE_CLUB
+    ONE_DIAMOND = Contracts.ONE_DIAMOND
+    ONE_HEART = Contracts.ONE_HEART
+    ONE_SPADE = Contracts.ONE_SPADE
+    ONE_NO_TRUMP = Contracts.ONE_NO_TRUMP
+    TWO_CLUBS = Contracts.TWO_CLUBS
+    TWO_DIAMONDS = Contracts.TWO_DIAMONDS
+    TWO_HEARTS = Contracts.TWO_HEARTS
+    TWO_SPADES = Contracts.TWO_SPADES
+    TWO_NO_TRUMP = Contracts.TWO_NO_TRUMP
+    THREE_CLUBS = Contracts.THREE_CLUBS
+    THREE_DIAMONDS = Contracts.THREE_DIAMONDS
+    THREE_HEARTS = Contracts.THREE_HEARTS
+    THREE_SPADES = Contracts.THREE_SPADES
+    THREE_NO_TRUMP = Contracts.THREE_NO_TRUMP
+    FOUR_CLUBS = Contracts.FOUR_CLUBS
+    FOUR_DIAMONDS = Contracts.FOUR_DIAMONDS
+    FOUR_HEARTS = Contracts.FOUR_HEARTS
+    FOUR_SPADES = Contracts.FOUR_SPADES
+    FOUR_NO_TRUMP = Contracts.FOUR_NO_TRUMP
+    FIVE_CLUBS = Contracts.FIVE_CLUBS
+    FIVE_DIAMONDS = Contracts.FIVE_DIAMONDS
+    FIVE_HEARTS = Contracts.FIVE_HEARTS
+    FIVE_SPADES = Contracts.FIVE_SPADES
+    FIVE_NO_TRUMP = Contracts.FIVE_NO_TRUMP
+    SIX_CLUBS = Contracts.SIX_CLUBS
+    SIX_DIAMONDS = Contracts.SIX_DIAMONDS
+    SIX_HEARTS = Contracts.SIX_HEARTS
+    SIX_SPADES = Contracts.SIX_SPADES
+    SIX_NO_TRUMP = Contracts.SIX_NO_TRUMP
+    SEVEN_CLUBS = Contracts.SEVEN_CLUBS
+    SEVEN_DIAMONDS = Contracts.SEVEN_DIAMONDS
+    SEVEN_HEARTS = Contracts.SEVEN_HEARTS
+    SEVEN_SPADES = Contracts.SEVEN_SPADES
+    SEVEN_NO_TRUMP = Contracts.SEVEN_NO_TRUMP
+    PASS = "PASS"
+    DOUBLE = Doubles.DOUBLE
+    REDOUBLE = Doubles.REDOUBLE
+
+    @staticmethod
+    def bids():
+        return [
+            Bids.ONE_CLUB,      Bids.ONE_DIAMOND,       Bids.ONE_HEART,     Bids.ONE_SPADE,     Bids.ONE_NO_TRUMP,
+            Bids.TWO_CLUBS,     Bids.TWO_DIAMONDS,      Bids.TWO_HEARTS,    Bids.TWO_SPADES,    Bids.TWO_NO_TRUMP,
+            Bids.THREE_CLUBS,   Bids.THREE_DIAMONDS,    Bids.THREE_HEARTS,  Bids.THREE_SPADES,  Bids.THREE_NO_TRUMP,
+            Bids.FOUR_CLUBS,    Bids.FOUR_DIAMONDS,     Bids.FOUR_HEARTS,   Bids.FOUR_SPADES,   Bids.FOUR_NO_TRUMP,
+            Bids.FIVE_CLUBS,    Bids.FIVE_DIAMONDS,     Bids.FIVE_HEARTS,   Bids.FIVE_SPADES,   Bids.FIVE_NO_TRUMP,
+            Bids.SIX_CLUBS,     Bids.SIX_DIAMONDS,      Bids.SIX_HEARTS,    Bids.SIX_SPADES,    Bids.SIX_NO_TRUMP,
+            Bids.SEVEN_CLUBS,   Bids.SEVEN_DIAMONDS,    Bids.SEVEN_HEARTS,  Bids.SEVEN_SPADES,  Bids.SEVEN_NO_TRUMP,
+            Bids.PASS,          Bids.DOUBLE,            Bids.REDOUBLE
+        ]
+
+    def is_sufficient_bid(self, by_bidder, current_contract):
```

Comment:
This encapsulates all the error checking, thus eliminating the need for the functions `pass_bid`, `make_bid`, `double`, and `redouble`. All we need now is one function called `get_new_bid` that calls this function.